Sources/OpenAPIRuntime/URICoder/Decoding/URIValueFromNodeDecoder.swift

```diff
@@ -167,153 +135,240 @@ extension URIValueFromNodeDecoder {
         throw DecodingError.typeMismatch(String.self, .init(codingPath: codingPath, debugDescription: message))
     }
 
-    /// Extracts the root value of the provided node using the root key.
-    /// - Parameter node: The node which to expect for the root key.
-    /// - Returns: The value found at the root key in the provided node.
-    /// - Throws: A `DecodingError` if the value is not found at the root key
-    private func rootValue(in node: URIParsedNode) throws -> URIParsedValueArray {
-        guard let value = node[rootKey] else {
-            if style == .simple, let valueForFallbackKey = node[""] {
-                // The simple style doesn't encode the key, so single values
-                // get encoded as a value only, and parsed under the empty
-                // string key.
-                return valueForFallbackKey
+    // MARK: - withParsed methods
+
+    /// Use the root as a primitive value.
+    /// - Parameter work: The closure in which to use the value.
+    /// - Returns: Any value returned from the closure.
+    /// - Throws: When parsing the root fails.
+    private func withParsedRootAsPrimitive<R>(_ work: (URIDecodedPrimitive?) throws -> R) throws -> R {
+        let value: URIDecodedPrimitive?
+        if let cached = cache.primitive {
+            value = try cached.get()
+        } else {
+            let result: Result<URIDecodedPrimitive?, any Error>
+            do {
+                value = try parser.parseRootAsPrimitive(rootKey: rootKey)?.value
+                result = .success(value)
+            } catch {
+                result = .failure(error)
+                throw error
             }
-            return []
+            cache.primitive = result
         }
-        return value
+        return try work(value)
```

Comment:
I'd like to see this pattern factored out as it's pretty repetitive.

I'm also wondering if `Optional<Result<>>` is the best representation for the state of the cache. I understand it fits with the three cases: (1) haven't tried, (2) tried and got a thing, and (3) tried and got an error; but I think an actual `enum CacheResult` with the three states in one enum might help us build some better layered logic.

Reply:
Factored the caching logic out and further simplified.

I kept the optional result type, as it gives us the nice `value.get()` "replay" ability of the cached value.